### DISCUSSION TITLE
Issues with Controller input state

### DISCUSSION THREAD

**Discussion by wdfinch** [react-hook-form/react-hook-form#3899] [Q&A] [answered]
Hi there, I'm having some issues regarding the input state passed from a `<Controller/>` to a component via the render prop. I'm not sure if this is a bug or I am just missing something here so I would really appreciate any assistance.

My issue is that the field state argument that is provided to the render function of a `<Controller/>` component does not seem to have working validation and is mismatched with the Typescript definition. 

Specifically, the state argument is always an object equal to `{invalid: false}` regardless of whether the validation passes or not. I have tried using both yup and rules for validation and neither changes the result.

Regarding the Typescript issue, the input state is defined as 

`export declare type InputState = {
    invalid: boolean;
    isTouched: boolean;
    isDirty: boolean;
};`

And what I receive does not match this (`isTouched` and `isDirty` are `undefined`).

Lastly, I am confused regarding how I can access a custom error message. I understand how to add an error message via yup or the library, but I don't understand how I can receive this into my component. In this [issue](https://github.com/react-hook-form/react-hook-form/issues/3434), @bluebill1049 mentions this: 

> ```tsx
> <Controller 
>   render={(props, { isDirty, isValid, error, isTouched } ) => {
>     console.log(state)
>   }}
> />
> `

However, it seems that the type definition file does not mention an error property (although I think this is a great solution).

I've included a code sandbox [here](https://codesandbox.io/s/musing-hugle-r8if6?file=/src/App.js) to show my exact setup and what I am receiving from the input state.

Thank you so much and really great work to everyone who's worked on this library. I can tell a ton of time and thought went into this and I am excited to be able to use this in my project.

react-hook-form: 6.14.1
yup: 0.32.8
@hookform/resolvers: 1.3.2

**Comment by bluebill1049** [comment]
> And what I receive does not match this (isTouched and isDirty are undefined)

That's a bug, can you reproduce this in a codesandbox?

> render={(props, { isDirty, isValid, error, isTouched } ) => {

instead of adding more states in the arg, we will have a new custom hook coming in v7, and it allows users to subscribe individual state and only re-render at the component, but it's not possible with the current v6 architecture. you will have to useFormContext and read errors object

**Reply by wdfinch** [reply]
Thanks for the reply. Yes, I attached a link above to my code sandbox. Here is it again:

https://codesandbox.io/s/musing-hugle-r8if6?file=/src/App.js

Sounds good. Any ideas on when this will be available? Also, do you have any feedback on why my validation isn't working in the attached sandbox?

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/fast-cherry-qr6db

don't mix schema with build-in, if you need to subscribe isValid, `mode` should be 'onChange'.

**Reply by wdfinch** [reply]
Ok so it seems like logging out `state.isDirty` and `state.isTouched` worked, but I'm not sure why when I log out the entire state object, those aren't present. It still looks like invalid is always set to false though. Additionally, the errors object seems to always be empty.

https://codesandbox.io/s/jovial-sun-w19kv?file=/src/App.js

Just added some additional logs.

**Comment by wdfinch** [comment, marked as answer]
The solution for anyone else who comes across this is to use the `mode` option on `useForm` to adjust the validation strategy.
